In [1]:
import Space,Generate

In [2]:
S = Generate.Gen(Generate.codas,3).space()

In [3]:
Even = Generate.EvenAtoms(2)

In [4]:
SE = S.apply(Even)

In [5]:
SEE = Space.multiEval(SE,16) 

In [6]:
SEE

eval: True, true: 5541, false:13792, undecided:3803, codas:23136, datas:23136

In [7]:
len(S),len(SE),len(SEE)

(11568, 23136, 23136)

In [8]:
for i in range(10):
    print(str(SEE[i]),'->',SEE._evals[SEE[i]])

({B aps} B:) -> aps
({B aps} B:◎ ◎) -> ◎ ◎ aps
({B ap} ap2:) -> ap
({B ap} ap2:◎ ◎) -> ◎ ◎ ap
({app B} collect:) -> app
({app B} collect:◎ ◎) -> app ◎ ◎
(apby {B none}:) -> 
(apby {B none}:◎ ◎) -> (apby {B none}:◎ ◎)
(tail A:) -> 
(tail A:◎ ◎) -> (tail A:◎ ◎)


In [9]:
Categories = S.category(Even,Even)

In [ ]:
len(Categories)

In [ ]:
CatE = Space.multiEval(Categories,16)

In [ ]:
CatE

In [11]:
R100 = Space.Space(*S.data_sample(100))

In [12]:
R100

eval: False, true: 0, false:0, undecided:0, codas:100, datas:100

In [13]:
Test = Space.Space(*[r for r in R100]+[e for e in Even])

In [14]:
Test

eval: False, true: 0, false:0, undecided:0, codas:100, datas:102

In [15]:
Categories2 = S.category(Test,Even)

In [16]:
Categories2

eval: False, true: 0, false:0, undecided:0, codas:2359872, datas:2359872

In [17]:
CatE = Space.multiEval(Categories2,16)

In [18]:
CatE

eval: True, true: 460444, false:1449048, undecided:450380, codas:2359872, datas:2359872

In [19]:
print(CatE[1000])

(= (tail A:rep 𝝞):(tail A:(tail A:rep 𝝞) (tail A:)))


In [20]:
CatEtrue = CatE.true()

In [21]:
CatEtrue

eval: False, true: 0, false:0, undecided:0, codas:460444, datas:460444

In [34]:
str(CatEtrue[1000][0].left()[-1].left())

'null if'

In [36]:
CatCan = Space.Space(*[d[0].left()[-1].left() for d in CatEtrue])

In [37]:
CatCan

eval: False, true: 0, false:0, undecided:0, codas:143, datas:460444

In [174]:
Ran1 = Space.Space(*S.data_sample(5))
Ran2 = Space.Space(*S.data_sample(5))

In [173]:
str(CatCan.sample()[0])

'split {B =}'

In [175]:
CatE2 = CatCan.category(Ran1,Ran2)

In [176]:
CatE2

eval: False, true: 0, false:0, undecided:0, codas:109125, datas:11511100

In [177]:
CatE2E = Space.multiEval(CatE2,16)

In [178]:
CatE2E

eval: True, true: 52819, false:40832, undecided:15474, codas:109125, datas:11511100

In [179]:
for i in range(10):
    print(CatE2E[i])

(= (apby {B none}:a a {A B} startswith):(apby {B none}:(apby {B none}:a a) (apby {B none}:{A B} startswith)))
(= (apby {B none}:a a join {ap B}):(apby {B none}:(apby {B none}:a a) (apby {B none}:join {ap B})))
(= (apby {B none}:a a wrap {B 𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬}):(apby {B none}:(apby {B none}:a a) (apby {B none}:wrap {B 𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬})))
(= (apby {B none}:a a {equiv B} 𝝞):(apby {B none}:(apby {B none}:a a) (apby {B none}:{equiv B} 𝝞)))
(= (apby {B none}:a a collect 1):(apby {B none}:(apby {B none}:a a) (apby {B none}:collect 1)))
(= (apby {B none}:dom null {A B} startswith):(apby {B none}:(apby {B none}:dom null) (apby {B none}:{A B} startswith)))
(= (apby {B none}:dom null join {ap B}):(apby {B none}:(apby {B none}:dom null) (apby {B none}:join {ap B})))
(= (apby {B none}:dom null wrap {B 𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬}):(apby {B none}:(apby {B none}:dom null) (apby {B none}:wrap {B 𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬})))
(= (apby {B none}:dom null {equiv B} 𝝞):(apby {B none}:(apby {B none}:dom null) (apby {B none}

In [180]:
CatE2E_true = CatE2E.true()

In [181]:
CatE2E_true

eval: False, true: 0, false:0, undecided:0, codas:52819, datas:52819

In [191]:
str(CatE2E_true[1])

'(= (imply {put B}:a a join {ap B}):(imply {put B}:(imply {put B}:a a) (imply {put B}:join {ap B})))'

In [196]:
str(CatE2E_true[1][0].left()[1].left())

'imply {put B}'

In [199]:
Candidates = set([])
for i in range(len(CatE2E_true)): Candidates.add(CatE2E_true[i][0].left()[1].left())

In [200]:
len(Candidates)

2186

In [202]:
S

eval: False, true: 0, false:0, undecided:0, codas:143, datas:11568

In [205]:
CandidateSpace = Space.Space(*[c for c in Candidates])

In [208]:
CandidateSpace.eval(100)

eval: True, true: 0, false:2186, undecided:0, codas:143, datas:2186

At this point, I have 2186 candidate categories.  I'm suspecting that many of these are not interesting in the sense that C:X = () always.  Let's filter these out by requiring C:X false for some X.

In [209]:
CandidateNonzero = CandidateSpace.apply(Test)

In [210]:
CandidateNonzero

eval: False, true: 0, false:0, undecided:0, codas:222972, datas:222972

In [211]:
CandidateNonzeroE = Space.multiEval(CandidateNonzero,16)

In [212]:
CandidateNonzeroE

eval: True, true: 119960, false:102978, undecided:34, codas:222972, datas:222972

In [213]:
CandidateNonzeroE_false = CandidateNonzeroE.false()

In [214]:
CandidateNonzeroE_false

eval: False, true: 0, false:0, undecided:0, codas:102978, datas:102978

In [219]:
str(CandidateNonzeroE_false[1][0].left())

'rep none'

In [222]:
CandidateSet2 = set([d[0].left() for d in CandidateNonzeroE_false])

1227

In [224]:
for d in CandidateSet2: print(d)

rep none
pass {B rep}
rep {post B}
some {𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬 B}
{B} =
some pre
{B} last
{= B} none
{B} collect
or {ap B}
pass {B put}
apif or
split nth1
aps pre
pass {post B}
or {b B}
endswith {B 𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬}
{B =} B
some {imply B}
rep {𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝝞𝝞𝟬𝝞𝝞𝟬𝟬 B}
split {B tail}
rep {= B}
aps {imply B}
rep {B join}
split A
startswith {B pre}
some endswith
rep {B or}
rep {aps B}
endswith {B get1}
pass {𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝝞𝝞𝟬𝝞𝝞𝟬𝟬 B}
pass {B join}
aps endswith
or {B pass}
rep {B rev}
startswith {B tail}
or or
split {B counts}
some {a B}
endswith aps
{B} rev
rep 2
endswith split
split {apif B}
pass {B rev}
= collect
startswith A
app pre
aps {a B}
{B =} app
some {B =}
{= B} 2
or equiv
aps {B =}
startswith {B counts}
{B =} 1
rep {put B}
or {or B}
startswith wrap
some if
endswith 0
pass {endswith B}
{B} none
split {has1 B}
rep {A B}
pass {put B}
endswith join
aps if
{= B} tail
imply
pass {pre B}
pass {last B}
{= B} ap2
endswith startswith
{B =} =
{B =} last
{B =} or
some 𝟬
pass b
app {B =}
ap imply
some 

In [225]:
Candidates2 = Space.Space(*[c for c in CandidateSet2])

In [226]:
Candidates2

eval: False, true: 0, false:0, undecided:0, codas:143, datas:1227

In [227]:
Idempotents = Candidates2.idempotent(Test)

In [229]:
IdempotentsE = Space.multiEval(Idempotents,16)

In [230]:
IdempotentsE

eval: True, true: 121236, false:3877, undecided:41, codas:125154, datas:125154

In [235]:
str(IdempotentsE[1][0].right()[0].left())

'rep none'

In [236]:
IdempotentTrue = IdempotentsE.true()

In [239]:
IdempotentSet = set([d[0].right()[0].left() for d in IdempotentTrue])

In [240]:
len(IdempotentSet)

1227

In [241]:
for c in Candidates2: print(c)

rep none
pass {B rep}
rep {post B}
some {𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬 B}
{B} =
some pre
{B} last
{= B} none
{B} collect
or {ap B}
pass {B put}
apif or
split nth1
aps pre
pass {post B}
or {b B}
endswith {B 𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝟬}
{B =} B
some {imply B}
rep {𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝝞𝝞𝟬𝝞𝝞𝟬𝟬 B}
split {B tail}
rep {= B}
aps {imply B}
rep {B join}
split A
startswith {B pre}
some endswith
rep {B or}
rep {aps B}
endswith {B get1}
pass {𝝞𝝞𝝞𝟬𝝞𝟬𝝞𝝞𝝞𝝞𝝞𝝞𝟬𝝞𝝞𝟬𝟬 B}
pass {B join}
aps endswith
or {B pass}
rep {B rev}
startswith {B tail}
or or
split {B counts}
some {a B}
endswith aps
{B} rev
rep 2
endswith split
split {apif B}
pass {B rev}
= collect
startswith A
app pre
aps {a B}
{B =} app
some {B =}
{= B} 2
or equiv
aps {B =}
startswith {B counts}
{B =} 1
rep {put B}
or {or B}
startswith wrap
some if
endswith 0
pass {endswith B}
{B} none
split {has1 B}
rep {A B}
pass {put B}
endswith join
aps if
{= B} tail
imply
pass {pre B}
pass {last B}
{= B} ap2
endswith startswith
{B =} =
{B =} last
{B =} or
some 𝟬
pass b
app {B =}
ap imply
some 